### DISCUSSION TITLE
edge function causes weird error

### DISCUSSION THREAD

**Discussion by Weiggonen** [vercel/next.js#55113] [Help]
### Summary

I need to increase processing time because creating chat completion from open ai api tekes about 5minutes, but edge function causes weird error in my api route.js. 

ERROR: - error models/program.js (33:0) @ <unknown>
- error Cannot read properties of undefined (reading 'Program')
  31 | );
  32 | 
> 33 | const Program = models.Program || model("Program", ProgramSchema);
  34 | 
  35 | export default Program;

I also tried to increase maxDuration with vercel.json file but it didn't help, don't know why. It wen't trough deploying but doesn't affect to timeoout limit.  

I'm new in vercel environment so any help is welcome. 

### Additional information

```js
route.js file: 

import createChatCompletion from "@/functions/generate";
import Program from "@/models/program";
import { connectToDB } from "@/utils/database";


export const runtime = "edge";

export async function POST(request) {
  let body = "";
  const now = new Date();

  try {

    const formData = await request.formData();


    const prompt = formData.get("prompt");
    const userId = formData.get("userId");

    const dataResponse = await createChatCompletion(prompt);

    await connectToDB();
    await Program.create({
      date: now,
      program: dataResponse,
      userId: userId,
      completed: false,
    });

    return new Response(dataResponse, { status: 200 }); // Assuming a 200 status code for OK.
  } catch (err) {
    console.error("Error reading or parsing request body:", err);
    return new Response("Internal server error", { status: 500 });
  }
}
```

**Comment by wwwdepot** [comment]
Read the doc: https://nextjs.org/docs/app/api-reference/edge#unsupported-apis

Any code you import into the edge runtime is using Node.js APIs, it will not work.

Move the db codes into a standalone API endpoint, do a fetch to save data.

```js
  const dataResponse = await createChatCompletion(prompt);

  // add this line
  await fetch("https://YOUR_DOMAIN/api/createProgram", { method: "POST", body: ..., headers: ... });

  return new Response(dataResponse, { status: 200 });
```

If you think your issue is a BUG, go create a new issue.

**Reply by Weiggonen** [reply]
Thank you for your help and time! I did'n found that unsupported apis before... But now when I call createChatCompletion function from edge function api route it crashes. It works if the api route isn't edge function. I appreciate if you have time to take a look at it:

```
import { Configuration, OpenAIApi } from "openai";

const configuration = new Configuration({
  apiKey: process.env.GPT_API_KEY,
});

const openai = new OpenAIApi(configuration);

async function createChatCompletion(message) {
  console.log(message);
  try {
    const formattedMessage = {
      role: "user",
      content: message,
    };
    const formattedSystemMessage = {
      role: "system",
      content: `Some message`,
    };

    const response = await openai.createChatCompletion({
      model: "gpt-3.5-turbo-16k",
      messages: [formattedSystemMessage, formattedMessage],
      temperature: 1,
      max_tokens: 15555,
    });
    const reply = response.data.choices[0].message.content;
    console.log(response.data.choices[0].message.content);
    return reply;
  } catch (error) {
    console.error("Error getting response from OpenAI:", error.response.data);
    throw error;
  }
}

export default createChatCompletion;

```

**Reply by icyJoseph** [reply]
Maybe you should read through this? https://www.npmjs.com/package/openai-edge not necessarily install it yet, but probably means that the openai kit doesn't work on Edge Env out of the box.

**Reply by Weiggonen** [reply]
Thank you I will check that.

**Reply by icyJoseph** [reply]
Btw, from the docs: https://nextjs.org/docs/app/building-your-application/routing/route-handlers#streaming they do use `openai-edge`
